Review

rithu1998 opened the pull request:
Pull request for code review branch for mini features.

arshana committed:
Initial push with no changes

arshana committed:
One more no-change push

Commit:
Added ServerInfo class

Commit:
Finished Up Time mini feature

sanjananatraj committed:
first build of tokenizing input mini feature

Commit:
Added server info class

Commit:
Added code to support uptime feature

Commit:
Finished up time mini feature

sanjananatraj committed:
fixed some errors with Chat.java

sanjananatraj committed:
Merge pull request #3 from arshana/sanjana-tokenizing-input

Sanjana tokenizing input

rithu1998 committed:
Merge branch 'review' into rithu-uptime-check

rithu1998 committed:
Merge pull request #2 from arshana/rithu-uptime-check

Rithu uptime check

sanjananatraj committed:
added import line for chat.java

arshana committed:
Version check activity

arshana committed:
Merge branch 'review' into arshana-version-check

arshana committed:
Merge pull request #5 from arshana/arshana-version-check

Version check activity

arshana committed:
Update Chat.java

arshana committed:
Added some comments

sanjananatraj committed:
ignored bin

sanjananatraj committed:
ignored bin

sanjananatraj committed:
removed the ignored directory bin

sanjananatraj committed:
removed chat logs

sanjananatraj committed:
tokenizing input should work nowgit status

sanjananatraj committed:
Merge branch 'review' into sanjana-tokenizing-input

sanjananatraj committed:
Merge pull request #6 from arshana/sanjana-tokenizing-input

Sanjana tokenizing input

sanjananatraj committed:
updated .gitignore to ignore itself and more chat logs

mbassiouny commented on .gitignore:
```diff
@@ -1,2 +1,13 @@
+.idea/
```
You do not need for that. I think commit https://github.com/google/codeu_project_2017/commit/ae0dc089d7e30063369c3b0d6982990c94f03df9 already done what should be ignored. Adding all of that will not enable anyone to commit code in these files directories. For example, adding "test/" will not enable anyone to write test code.

arshana replied:
I've fixed this.

mbassiouny commented on README.md:
```diff
@@ -125,3 +125,6 @@ Classes that are shared by the clients and servers.
 ### codeu.chat.util
 
 Some basic infrastructure classes used throughout the project.
+
```
Why this bunch of new lines?

arshana replied:
I've fixed this.

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -18,11 +18,18 @@
 import java.util.Map;
 import java.util.Scanner;
 import java.util.Stack;
+import java.util.List;
```
Nit. Style should be alpha order.

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -18,11 +18,18 @@
+import java.util.ArrayList;
+import java.io.IOException;
+
+import codeu.chat.common.ServerInfo;
```
Nit: Style, this should go with the next set of imports.

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -18,11 +18,18 @@
 import codeu.chat.client.core.MessageContext;
 import codeu.chat.client.core.UserContext;
+import codeu.chat.util.Tokenizer;
+import codeu.chat.common.ServerInfo;
```
The same here, alpha order.

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -47,9 +54,18 @@ public Chat(Context context) {
-    final String command = tokens.hasNext() ? tokens.next() : "";
+    final List<String> args = new ArrayList<>();
+    final Tokenizer tokenizer = new Tokenizer(line);
+    try{
```
Nit, missing space after try, also missing spaces before the for loop.

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -136,8 +156,8 @@ public void invoke(Scanner args) {
-      public void invoke(Scanner args) {
-        final String name = args.hasNext() ? args.nextLine().trim() : "";
+      public void invoke(List<String> args) {
+        final String name = args.size() > 0 ? args.get(0) : "";
```
This is not necessary. You should check for args.size() == 1, otherwise print exception. Now you are putting empty string and checking for the empty string which does not seem natural.

sanjananatraj replied:
@mbassiouny 
can you elaborate on this? would
`final String name = args.size() == 1 ? args.get(0) : "";`
be okay? And instead of having the empty string, you want to print the exception instead, is that right?

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -155,8 +175,8 @@ public void invoke(Scanner args) {
-      public void invoke(Scanner args) {
-        final String name = args.hasNext() ? args.nextLine().trim() : "";
+      public void invoke(List<String> args) {
+        final String name = args.size() > 0 ? args.get(0) : "";
```
The same as the previous comment. There are other few places in the code, please fix them all.

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -180,6 +200,25 @@ private UserContext findUser(String name) {
       }
     });
+    
+    //Added this code to respond to the user's request of info
```
Please follow the same standard when defining a new command. Example:

     // U-SIGN-IN (sign in user)
     //
     // Add a command to sign-in as a user when the user enters "u-sign-in"
     // while on the root panel.

arshana replied:
I've fixed this.

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -238,8 +277,8 @@ public void invoke(Scanner args) {
-      public void invoke(Scanner args) {
-        final String name = args.hasNext() ? args.nextLine().trim() : "";
+      public void invoke(List<String> args) {
+        final String name = args.size() > 0 ? args.get(0) : "";
```
see above.

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -260,8 +299,8 @@ public void invoke(Scanner args) {
-      public void invoke(Scanner args) {
-        final String name = args.hasNext() ? args.nextLine().trim() : "";
+      public void invoke(List<String> args) {
+        final String name = args.size() > 0 ? args.get(0) : "";
```
see above.

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -361,8 +400,8 @@ public void invoke(Scanner args) {
-      public void invoke(Scanner args) {
-        final String message = args.hasNext() ? args.nextLine().trim() : "";
+      public void invoke(List<String> args) {
+        final String message = args.size() > 0 ? args.get(0) : "";
```
see above.

mbassiouny commented on src/codeu/chat/client/commandline/Panel.java:
```diff
@@ -17,6 +17,7 @@
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Scanner;
+import java.util.List;
```
Nit: Order please.

arshana replied:
I've fixed this.

mbassiouny commented on src/codeu/chat/client/core/Context.java:
```diff
@@ -19,13 +19,14 @@
 public final class Context {
 
-  private final BasicView view;
+  private final View view;
```
Why you are changing that, that sounds to me out of context.

arshana replied:
@mbassiouny 
We changed this because at the bottom of the class, we added a method called getInfo(). It calls the view.getInfo() method. That getInfo() method isn't in the BasicView class, but it is in the View class, which inherits from BasicView. Rithu and I had both decided to go with the option of changing the variable view's type to View instead of adding the getInfo() method to BasicView. Should we add the method to BasicView instead of changing view's type?

mbassiouny replied:
You have to define the method in the interface "BasicView", and you already implemented it in View. Now you are making this class dependent of a specific implementation(View). The View Class is responsible for the network calls, consider that you want to implement a different ViewV2 class that uses a different code to do the network calls, now this class won't work because you make it dependent on a specific implementation.

arshana replied:
I've fixed this.

mbassiouny commented on src/codeu/chat/client/core/Context.java:
```diff
@@ -47,4 +48,10 @@ public UserContext create(String name) {
+  //Added during Version Check technical activity
+  //Added this method to get the info from the server
+  public ServerInfo getInfo() {
+	  return view.getInfo();
```
Nit: Number of spaces.

mbassiouny commented on src/codeu/chat/common/NetworkCode.java:
```diff
@@ -35,5 +35,8 @@
       RELAY_WRITE_REQUEST = 29,
-      RELAY_WRITE_RESPONSE = 30;
+      RELAY_WRITE_RESPONSE = 30,
+      //Next two variables defined during Version Check technical activity
```
Comments are not needed in here. It's self explaining itself.

arshana replied:
I've fixed this.

mbassiouny commented on src/codeu/chat/client/core/View.java:
```diff
@@ -17,18 +17,14 @@
-import codeu.chat.common.Message;
-import codeu.chat.common.NetworkCode;
-import codeu.chat.common.User;
+import codeu.chat.common.*;
```
It's better to use individual import so you know what you are importing.

rithu1998 replied:
Fixed this problem

mbassiouny commented on src/codeu/chat/client/core/View.java:
```diff
@@ -17,18 +17,14 @@
 import codeu.chat.util.Uuid;
 import codeu.chat.util.connections.Connection;
 import codeu.chat.util.connections.ConnectionSource;
+import sun.nio.ch.Net;
```
Why you need this import?

rithu1998 replied:
Took this import out

mbassiouny commented on src/codeu/chat/client/core/View.java:
```diff
@@ -117,23 +113,65 @@ public View(ConnectionSource source) {
   public Collection<Message> getMessages(Collection<Uuid> ids) {
 
-    final Collection<Message> messages = new ArrayList<>();
+      final Collection<Message> messages = new ArrayList<>();
```
nit: formatting issue. please follow the files format for the indentation.

mbassiouny replied:
Also the remaining of the file does not follow the indentation scheme, please fix all.

arshana replied:
I've fixed this.

mbassiouny commented on src/codeu/chat/client/core/View.java:
```diff
@@ -117,23 +113,65 @@ public View(ConnectionSource source) {
+      return null;
   }
+  
+  public ServerInfo getInfo() {
```
Is that a duplicate?

mbassiouny replied:
That might be a result of a merge conflict. Please remove.

arshana replied:
@mbassiouny 
The differences between the two getInfo() methods is on lines 147-148 and lines 168-169.
I'm not sure how to resolve this issue. Line 168 can just be moved up to the first getInfo() method. But line 148 and line 169 are both return lines. I'm not sure how to deal with that. What advice do you have?

mbassiouny replied:
May be you add a constructor of server info that takes version and start time, and then construct the serverinfo object using the version and start time. By that you can combine the 2 pieces of code together.

arshana replied:
I've fixed this.

mbassiouny commented on src/codeu/chat/client/core/View.java:
```diff
@@ -117,23 +113,65 @@ public View(ConnectionSource source) {
+      return messages;
+  }
+
+  //Added this method during Version Check technical activity.
```
Just state what the method is doing.

arshana replied:
I've fixed this.

mbassiouny commented on src/codeu/chat/common/BasicView.java:
```diff
@@ -51,5 +51,4 @@
   //
   //   Return all messages whose id is found in the given collection.
   Collection<Message> getMessages(Collection<Uuid> ids);
-
```
Not related change.

arshana replied:
I've fixed this.

mbassiouny commented on src/codeu/chat/common/ServerInfo.java:
```diff
@@ -0,0 +1,40 @@
+package codeu.chat.common;
```
Follow the other files standard with the import order.

arshana replied:
I've fixed this.

mbassiouny commented on src/codeu/chat/common/ServerInfo.java:
```diff
@@ -0,0 +1,40 @@
+ * Holds Server's information
+ * Used by both Client and Server
+ */
+public final class ServerInfo {
```
Please follow the indentation scheme for the other files.

arshana replied:
I've fixed this.

mbassiouny commented on src/codeu/chat/common/ServerInfo.java:
```diff
@@ -0,0 +1,40 @@
+package codeu.chat.common;
```
I do not see unit tests for this class, please add.

mbassiouny commented on src/codeu/chat/server/Server.java:
```diff
@@ -25,14 +25,8 @@
-import codeu.chat.common.Relay;
-import codeu.chat.common.Secret;
-import codeu.chat.common.User;
+import codeu.chat.common.*;
```
Naming each one is better.

mbassiouny commented on src/codeu/chat/server/Server.java:
```diff
@@ -63,6 +57,11 @@
+  /* This was added to create a ServerInfo object*/
+  private static final ServerInfo info = new ServerInfo();
+
+  //added during Version Check technical activity
```
Sounds like a merge conflict in here. Please remove.

arshana replied:
I've fixed this.

mbassiouny commented on src/codeu/chat/server/Server.java:
```diff
@@ -63,6 +57,11 @@
 
   private final Relay relay;
   private Uuid lastSeen = Uuid.NULL;
+  /* This was added to create a ServerInfo object*/
```
Nit: Usually in big projects with big team member, no one will follow when this added. Generally you have to document what is this variable, and in what way it used if it's not obvious.

arshana replied:
I've removed the comment.

mbassiouny commented on src/codeu/chat/server/Server.java:
```diff
@@ -171,6 +170,15 @@ public void onMessage(InputStream in, OutputStream out) throws IOException {
     });
+    
+    //Added this code to handle server requests
+    this.commands.put(NetworkCode.SERVER_INFO_REQUEST, new Command() {
```
Nit: indentation.

arshana replied:
I've fixed this.

mbassiouny commented on src/codeu/chat/server/Server.java:
```diff
@@ -205,6 +213,18 @@ public void run() {
+        	  Time.SERIALIZER.write(connection.out(), info.startTime);
+          }
+
+          //Added this if statement during Version Check technical activity.
```
Another merge conflict.

arshana replied:
I've fixed this.

mbassiouny commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,75 @@
+package codeu.chat.util;
+import java.io.IOException;
```
May be add an empty line.

mbassiouny commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,75 @@
+import java.io.IOException;
+
+public final class Tokenizer{
+	private StringBuilder token;
```
Indentation is not correct. Please follow the other files indentation scheme.

mbassiouny commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,75 @@
+
+
+	
+} //end class
```
Add new line in the end of the file.

mbassiouny commented on test/codeu/chat/TestRunner.java:
```diff
@@ -28,7 +28,8 @@ public static void main(String[] args) {
              codeu.chat.util.UuidTest.class,
-             codeu.chat.util.store.StoreTest.class
+             codeu.chat.util.store.StoreTest.class,
+		  codeu.chat.util.TokenizerTest.class	
```
Indentation is not matching.

mbassiouny reviewed (changes requested):
Thanks all for your changes. Please see my comments, please do not get overwhelmed, may be this is the first professional code review, so that's why it might have several comments. But I think all of them are minor. I have more general comments:
1. Please do not check in .DS_Store files, that sounds like OS specific files.
2. I see you have 23 commits for 3 feature. This pollute the history and it's going to be hard to track. The optimal to see 1 commit/feature with total 3 commits. If that's hard to adjust, may be we can squash all of them to 1 when merging into summer-2017.

arshana committed:
Edits after code review

arshana committed:
Updates after Code Review

arshana committed:
Updated after Code Review

arshana committed:
Edited after code review

arshana committed:
Edits after Code review

arshana commented on src/codeu/chat/client/core/Context.java:
```diff
@@ -19,13 +19,14 @@
 public final class Context {
 
-  private final BasicView view;
+  private final View view;
```
We changed this because at the bottom of the class, we added a method called getInfo(). It calls the view.getInfo() method. That getInfo() method isn't in the BasicView class, but it is in the View class, which inherits from BasicView. Rithu and I had both decided to go with the option of changing the variable view's type to View instead of adding the getInfo() method to BasicView. Should we add the method to BasicView instead of changing view's type?

rithu1998 committed:
Update Server.java

Added back the individual imports

rithu1998 committed:
Update Chat.java

rithu1998 committed:
Update Context.java

rithu1998 committed:
Update View.java

sanjananatraj committed:
applied changes to Chat.java/Tokenizer.java/TestRunner.java

sanjananatraj committed:
updated Chat.java

sanjananatraj committed:
updated Chat.java + merged my code

sanjananatraj committed:
chat.java spacing issues

sanjananatraj committed:
Update TestRunner.java

I updated the spacing issue locally and pushed it, but it didn't seem to transfer over to GitHub, so I'm manually doing it here.

sanjananatraj committed:
Merge branch 'review' into sanjana-tokenizing-input

sanjananatraj committed:
Merge pull request #8 from arshana/sanjana-tokenizing-input

Sanjana tokenizing input

arshana committed:
Updated View.java after code review

arshana committed:
Update View.java

arshana committed:
Update View.java

arshana committed:
Update BasicView.java

arshana committed:
Update ServerInfo.java

arshana committed:
Update View.java

arshana committed:
Update BasicView.java

arshana committed:
Update View.java

arshana committed:
Update Context.java

arshana committed:
Update .gitignore

arshana committed:
Update Chat.java

arshana committed:
Update ServerInfo.java

Commit:
Final changes to mini features

Commit:
Small change to ServerInfoTest

arshana committed:
Delete .DS_Store

arshana committed:
Delete .DS_Store

arshana committed:
Delete .DS_Store

arshana committed:
Delete .DS_Store

sanjananatraj committed:
changed Chat.java

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -155,17 +176,17 @@ public void invoke(Scanner args) {
-        if (name.length() > 0) {
+      public void invoke(List<String> args) {
+        if(args.size() != 1){
+          System.out.println("ERROR: Failed to add new user");
```
We are not adding new user in here. Here we are signing in. May be the error message something like "Missing usename" or something.

sanjananatraj replied:
Fixed

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -155,17 +176,17 @@ public void invoke(Scanner args) {
           final UserContext user = findUser(name);
           if (user == null) {
             System.out.format("ERROR: Failed to sign in as '%s'\n", name);
-          } else {
```
We still need this else. Why you are removing it. We will do panels.push(createUserPanel(user)); only if the user is not null.

sanjananatraj replied:
Fixed

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -238,17 +280,17 @@ public void invoke(Scanner args) {
-        if (name.length() > 0) {
+      public void invoke(List<String> args) {
+        if(args.size() != 1){
+          System.out.println("ERROR: Failed to add new user");
```
We are not adding users in here. I think we are expecting the title.

sanjananatraj replied:
@mbassiouny Can you clarify this comment? Expecting the title of what, the conversation? Would something like "Failed to title new conversation" work or something else?

sanjananatraj replied:
I've put "Failed to title new conversation" for now, please tell me if I should change it!

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -238,17 +280,17 @@ public void invoke(Scanner args) {
           }
-        } else {
-          System.out.println("ERROR: Missing <title>");
+            panels.push(createConversationPanel(conversation));
```
We still need else in here, we won't create the panel unless the conversation is not null.

sanjananatraj replied:
Fixed

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -260,18 +302,18 @@ public void invoke(Scanner args) {
-        if (name.length() > 0) {
+      public void invoke(List<String> args) {
+        if(args.size() != 1){
+          System.out.println("ERROR: Failed to add new user");
```
The same as the previous 2 comments.

sanjananatraj replied:
fixed

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -361,12 +403,16 @@ public void invoke(Scanner args) {
-        } else {
+      public void invoke(List<String> args) {
+        if(args.size() != 1){
+          System.out.println("ERROR: Failed to add new user");
```
This is adding messages not users.

sanjananatraj replied:
fixed

mbassiouny commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -361,12 +403,16 @@ public void invoke(Scanner args) {
+        }
+        else{
+          String message = args.get(0);
+          if (message.length() < 0) {
```
This if is not needed, we sure that args.get(0) will have something because we are using the tokenizer which ensures that this won't be empty.

sanjananatraj replied:
@mbassiouny Just to clarify, we don't need the if statement within this else statement, correct? so having it as
`else{
 String message = args.get(0);
conversation.add(message);`
will work?

sanjananatraj replied:
removed if statement, please tell me if I should change that!

mbassiouny reviewed (changes requested):
That @rithu1998, please take a look on that. Hopefully this will be the final iteration.

sanjananatraj committed:
fixed Chat.java
